File: android/app/src/main/java/com/slackoff/app/support/BookshelfStore.kt
package com.slackoff.app.support

import android.content.Context
import android.content.SharedPreferences
import com.slackoff.app.model.NovelBook
import com.slackoff.app.model.NovelBookDetail
import com.slackoff.app.model.novelKey
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

/**
 * 书架上的一本书 + 精确阅读进度（章节 id / 章节序号 / 章内位置 / 全书百分比）。
 * 章内位置让「继续阅读」能回到上次那一页，而不只是那一章。
 */
@Serializable
data class ShelfBook(
    val source: String,
    val bookId: String,
    val title: String,
    val author: String? = null,
    val category: String? = null,
    val lastChapterId: String = "",
    val lastChapterTitle: String = "",
    /** 1-based 章节序号，与目录里的 index 对齐 */
    val lastChapterIndex: Int = 0,
    /** 目录总章数，用于算全书百分比；未知时为 0 */
    val totalChapters: Int = 0,
    /** 章内位置 0~1 */
    val positionInChapter: Double = 0.0,
    val updatedAt: Long = 0L,
) {
    val key: String get() = novelKey(source, bookId)

    /** 全书进度 0~1；总章数未知时退化为章内位置。 */
    val percent: Double
        get() {
            if (totalChapters <= 0) return positionInChapter
            val done = maxOf(lastChapterIndex - 1, 0).toDouble() + positionInChapter
            return (done / totalChapters).coerceIn(0.0, 1.0)
        }

    // 四舍五入，与 iOS 端 ShelfBook.percentText 保持一致：
    // 截断取整会让 9.95% 显示成 "9%"，两端同一本书进度显示不同。
    val percentText: String get() = "${kotlin.math.round(percent * 100).toInt()}%"

    val progressText: String
        get() =
            if (lastChapterIndex > 0) {
                "读至 第${lastChapterIndex}章"
            } else if (lastChapterTitle.isEmpty()) {
                "尚未开始"
            } else {
                "读至 $lastChapterTitle"
            }
}

/**
 * 书架与阅读进度的本地存储（SharedPreferences + StateFlow）。
 * 与 iOS 的 BookshelfStore 对齐：同一份语义、同一套 key 规则。
 */
object BookshelfStore {
    private const val PREFS_NAME = "com.slackoff.novel-shelf"
    private const val KEY = "books"

    private val json = Json {
        ignoreUnknownKeys = true
        isLenient = true
        explicitNulls = false
        encodeDefaults = true
    }

    private var prefs: SharedPreferences? = null
    private val _books = MutableStateFlow<List<ShelfBook>>(emptyList())

    /** 已按最后阅读时间倒序。 */
    val books: StateFlow<List<ShelfBook>> = _books.asStateFlow()

    /** 「继续阅读」卡片用：最近读过的那本。 */
    val mostRecent: ShelfBook? get() = _books.value.firstOrNull()

    fun init(context: Context) {
        if (prefs != null) return
        val p = context.applicationContext.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        prefs = p
        val raw = p.getString(KEY, null) ?: return
        _books.value = decode(raw)
    }

    internal fun decode(raw: String): List<ShelfBook> {
        return try {
            json.decodeFromString<List<ShelfBook>>(raw).sortedByDescending { it.updatedAt }
        } catch (e: Exception) {
            emptyList()
        }
    }

    fun contains(key: String): Boolean = _books.value.any { it.key == key }

    fun contains(source: String, bookId: String): Boolean = contains(novelKey(source, bookId))

    fun book(key: String): ShelfBook? = _books.value.firstOrNull { it.key == key }

    fun book(source: String, bookId: String): ShelfBook? = book(novelKey(source, bookId))

    fun add(book: NovelBook) {
        add(
            source = book.source,
            bookId = book.bookId,
            title = book.title,
            author = book.author,
            category = book.category,
            chapterId = book.latestChapterId.orEmpty(),
            chapterTitle = book.latestChapter.orEmpty(),
            chapterIndex = 0,
        )
    }

    fun add(detail: NovelBookDetail) {
        add(
            source = detail.source,
            bookId = detail.bookId,
            title = detail.title,
            author = detail.author,
            category = detail.category,
            chapterId = detail.entryChapterId.orEmpty(),
            chapterTitle = detail.firstChapter ?: detail.latestChapter.orEmpty(),
            chapterIndex = if (detail.entryChapterId == detail.firstChapterId) 1 else 0,
            totalChapters = detail.chapterCount ?: 0,
        )
    }

    /** 加入书架；已在架则只更新元信息（书名/作者可能被源站修正），不动进度。 */
    fun add(
        source: String,
        bookId: String,
        title: String,
        author: String? = null,
        category: String? = null,
        chapterId: String,
        chapterTitle: String,
        chapterIndex: Int,
        totalChapters: Int = 0,
    ) {
        val key = novelKey(source, bookId)
        val existing = book(key)
        val next =
            existing?.copy(
                title = title,
                author = author ?: existing.author,
                category = category ?: existing.category,
                totalChapters = if (totalChapters > 0) totalChapters else existing.totalChapters,
                updatedAt = System.currentTimeMillis(),
            ) ?: ShelfBook(
                source = source,
                bookId = bookId,
                title = title,
                author = author,
                category = category,
                lastChapterId = chapterId,
                lastChapterTitle = chapterTitle,
                lastChapterIndex = chapterIndex,
                totalChapters = totalChapters,
                updatedAt = System.currentTimeMillis(),
            )
        replace(next)
    }

    /** 更新阅读进度（翻页节流 + 退出阅读器时落盘）。 */
    fun updateProgress(
        source: String,
        bookId: String,
        title: String,
        author: String? = null,
        category: String? = null,
        chapterId: String,
        chapterTitle: String,
        chapterIndex: Int,
        totalChapters: Int,
        positionInChapter: Double,
    ) {
        val key = novelKey(source, bookId)
        val existing = book(key)
        val next =
            existing?.copy(
                title = title,
                author = author ?: existing.author,
                category = category ?: existing.category,
                lastChapterId = chapterId,
                lastChapterTitle = chapterTitle,
                lastChapterIndex = chapterIndex,
                totalChapters = if (totalChapters > 0) totalChapters else existing.totalChapters,
                positionInChapter = positionInChapter.coerceIn(0.0, 1.0),
                updatedAt = System.currentTimeMillis(),
            ) ?: ShelfBook(
                source = source,
                bookId = bookId,
                title = title,
                author = author,
                category = category,
                lastChapterId = chapterId,
                lastChapterTitle = chapterTitle,
                lastChapterIndex = chapterIndex,
                totalChapters = totalChapters,
                positionInChapter = positionInChapter.coerceIn(0.0, 1.0),
                updatedAt = System.currentTimeMillis(),
            )
        replace(next)
    }

    fun remove(key: String) {
        _books.value = _books.value.filterNot { it.key == key }
        persist()
    }

    /** 清空书架（设置页用）。 */
    fun removeAll() {
        _books.value = emptyList()
        persist()
    }

    /** 覆盖同 key 条目并重新按 updatedAt 倒序（保持「最近读过」在最前）。 */
    private fun replace(book: ShelfBook) {
        val current = _books.value.toMutableList()
        val index = current.indexOfFirst { it.key == book.key }
        if (index >= 0) current[index] = book else current.add(book)
        _books.value = current.sortedByDescending { it.updatedAt }
        persist()
    }

    private fun persist() {
        val p = prefs ?: return
        val encoded = try {
            json.encodeToString<List<ShelfBook>>(_books.value)
        } catch (e: Exception) {
            return
        }
        p.edit().putString(KEY, encoded).apply()
    }
}

File: SlackOff/Support/BookshelfStore.swift
import Foundation

/// 书架上的一本书 + 精确阅读进度。
///
/// 进度粒度：章节 id / 章节序号 / 章内位置（0~1，页序比例）/ 全书百分比。
/// 章内位置让「继续阅读」能精确回到上次那一页，而不只是那一章。
struct ShelfBook: Codable, Identifiable, Hashable {
    let source: String
    let bookId: String
    var title: String
    var author: String?
    var category: String?

    var lastChapterId: String
    var lastChapterTitle: String
    /// 1-based 章节序号，与目录里的 index 对齐
    var lastChapterIndex: Int
    /// 目录总章数，用于算全书百分比；未知时为 0
    var totalChapters: Int
    /// 章内位置 0~1
    var positionInChapter: Double
    var updatedAt: Date

    var id: String { NovelBook.key(source: source, bookId: bookId) }

    /// 全书进度 0~1。总章数未知时退化为「章内位置」。
    var percent: Double {
        guard totalChapters > 0 else { return positionInChapter }
        let done = Double(max(lastChapterIndex - 1, 0)) + positionInChapter
        return min(max(done / Double(totalChapters), 0), 1)
    }

    var percentText: String { "\(Int((percent * 100).rounded()))%" }

    /// 书架格子上的「读至 第x章」文案。
    var progressText: String {
        if lastChapterIndex > 0 { return "读至 第\(lastChapterIndex)章" }
        return lastChapterTitle.isEmpty ? "尚未开始" : "读至 \(lastChapterTitle)"
    }
}

/// 书架与阅读进度的本地存储（UserDefaults，与 EpisodeStore 同风格）。
///
/// 用 @Observable 而非纯静态 enum：书架页需要在阅读器退出后立刻刷新，
/// 书城/详情页的「已在书架」状态也要跟着变。
@Observable
final class BookshelfStore {
    static let shared = BookshelfStore()

    static let defaultStorageKey = "com.slackoff.novel-shelf"
    private let storageKey: String
    /// 可注入的 UserDefaults：单测用独立 suite，避免污染真机书架数据。
    private let defaults: UserDefaults
    private let encoder = JSONEncoder()
    private let decoder = JSONDecoder()

    private(set) var books: [ShelfBook] = []

    init(
        defaults: UserDefaults = .standard,
        storageKey: String = BookshelfStore.defaultStorageKey
    ) {
        self.defaults = defaults
        self.storageKey = storageKey
        guard
            let data = defaults.data(forKey: storageKey),
            let decoded = try? decoder.decode([ShelfBook].self, from: data)
        else { return }
        books = decoded.sorted { $0.updatedAt > $1.updatedAt }
    }

    // MARK: 查询

    /// 按最后阅读时间倒序（书架页直接用）。
    var sorted: [ShelfBook] { books }

    /// 「继续阅读」卡片用：最近读过的那本。
    var mostRecent: ShelfBook? { books.first }

    func contains(_ id: String) -> Bool {
        books.contains { $0.id == id }
    }

    func contains(source: String, bookId: String) -> Bool {
        contains(NovelBook.key(source: source, bookId: bookId))
    }

    func book(_ id: String) -> ShelfBook? {
        books.first { $0.id == id }
    }

    func book(source: String, bookId: String) -> ShelfBook? {
        book(NovelBook.key(source: source, bookId: bookId))
    }

    // MARK: 写入

    /// 加入书架；已在架则只更新元信息（书名/作者可能被源站修正），**不动进度**。
    @discardableResult
    func add(_ novel: NovelBook) -> ShelfBook {
        add(
            source: novel.source,
            bookId: novel.bookId,
            title: novel.title,
            author: novel.author,
            category: novel.category,
            chapterId: novel.latestChapterId ?? "",
            chapterTitle: novel.latestChapter ?? "",
            chapterIndex: 0
        )
    }

    @discardableResult
    func add(_ detail: NovelBookDetail) -> ShelfBook {
        add(
            source: detail.source,
            bookId: detail.bookId,
            title: detail.title,
            author: detail.author,
            category: detail.category,
            chapterId: detail.entryChapterId ?? "",
            chapterTitle: detail.firstChapter ?? detail.latestChapter ?? "",
            chapterIndex: detail.entryChapterId == detail.firstChapterId ? 1 : 0,
            totalChapters: detail.chapterCount ?? 0
        )
    }

    @discardableResult
    func add(
        source: String,
        bookId: String,
        title: String,
        author: String? = nil,
        category: String? = nil,
        chapterId: String,
        chapterTitle: String,
        chapterIndex: Int,
        totalChapters: Int = 0
    ) -> ShelfBook {
        let id = NovelBook.key(source: source, bookId: bookId)
        if let existing = book(id) {
            var updated = existing
            updated.title = title
            updated.author = author ?? existing.author
            updated.category = category ?? existing.category
            if totalChapters > 0 { updated.totalChapters = totalChapters }
            replace(updated)
            return updated
        }
        let new = ShelfBook(
            source: source,
            bookId: bookId,
            title: title,
            author: author,
            category: category,
            lastChapterId: chapterId,
            lastChapterTitle: chapterTitle,
            lastChapterIndex: chapterIndex,
            totalChapters: totalChapters,
            positionInChapter: 0,
            updatedAt: .now
        )
        books.insert(new, at: 0)
        persist()
        return new
    }

    /// 更新阅读进度（翻页节流调用 + 退出阅读器时落盘）。
    @discardableResult
    func updateProgress(
        source: String,
        bookId: String,
        title: String,
        author: String? = nil,
        category: String? = nil,
        chapterId: String,
        chapterTitle: String,
        chapterIndex: Int,
        totalChapters: Int,
        positionInChapter: Double
    ) -> ShelfBook {
        var target = book(source: source, bookId: bookId) ?? ShelfBook(
            source: source,
            bookId: bookId,
            title: title,
            author: author,
            category: category,
            lastChapterId: chapterId,
            lastChapterTitle: chapterTitle,
            lastChapterIndex: chapterIndex,
            totalChapters: totalChapters,
            positionInChapter: positionInChapter,
            updatedAt: .now
        )
        target.title = title
        if let author { target.author = author }
        if let category { target.category = category }
        target.lastChapterId = chapterId
        target.lastChapterTitle = chapterTitle
        target.lastChapterIndex = chapterIndex
        if totalChapters > 0 { target.totalChapters = totalChapters }
        target.positionInChapter = min(max(positionInChapter, 0), 1)
        target.updatedAt = .now
        replace(target)
        return target
    }

    func remove(_ id: String) {
        books.removeAll { $0.id == id }
        persist()
    }

    /// 清空书架（设置页用）。
    func removeAll() {
        books.removeAll()
        persist()
    }

    // MARK: 内部

    /// 覆盖同 id 条目并重新按 updatedAt 倒序（保持「最近读过」在最前）。
    private func replace(_ book: ShelfBook) {
        if let index = books.firstIndex(where: { $0.id == book.id }) {
            books[index] = book
        } else {
            books.append(book)
        }
        books.sort { $0.updatedAt > $1.updatedAt }
        persist()
    }

    private func persist() {
        guard let data = try? encoder.encode(books) else { return }
        defaults.set(data, forKey: storageKey)
    }
}
